bundles/org.eclipse.ui.ide/extensions/org/eclipse/ui/actions/WorkspaceModifyOperation.java

```diff
@@ -137,9 +134,12 @@ public synchronized final void run(IProgressMonitor monitor)
 			interruptedException.initCause(e);
 			throw interruptedException;
 		}
-		// Re-throw the InvocationTargetException, if any occurred
-		if (iteHolder[0] != null) {
-			throw iteHolder[0];
+
+		// Re-throw the InvocationTargetException, if any occurred inside the IWorkspaceRunnable
+		if (exceptionHolder[0] instanceof InvocationTargetException ite) {
+			throw ite;
+		} else if (exceptionHolder[0] instanceof InterruptedException ie) {
+			throw ie;
 		}
```

Comment:
How would you write it?

Reply:
Now I see, the problem is that the `exceptionHolder` was generalized to `Exception` and the method declares specific types of exception. But then I would propose to store the typed exceptions separately, i.e., in different variables, rather than re-extracting them using `instanceof`.

Reply:
Thank you for clarifying it but I'd rather keep it as it is now: adding more variables only adds complexity to the method and it doesn't really improve readability.

Reply:
Your choice to select some other solution, but the current proposal contains an incomplete instanceof switch, which will eventually break.

Reply:
Then I must have misunderstood your proposal with the 2 variables because the way I imagined it would have also ended up being an incomplete switch in disguise. 

Could you please provide a snippet?

Reply:
Something like:
```java
		AtomicReference<InvocationTargetException> rethrownInvocationTargetException = new AtomicReference<>();
		AtomicReference<InterruptedException> rethrownInterruptedException = new AtomicReference<>();
		try {
			IWorkspaceRunnable workspaceRunnable = pm -> {
				try {
					execute(pm);
				} catch (InvocationTargetException e1) {
					rethrownInvocationTargetException.set(e1);
				} catch (InterruptedException e2) {
					rethrownInterruptedException.set(e2);
				}
				// CoreException and OperationCanceledException are propagated
			};
			...
		} catch (CoreException e) {
			throw new InvocationTargetException(e);
		} catch (OperationCanceledException e) {
			InterruptedException interruptedException = new InterruptedException(e.getMessage());
			interruptedException.initCause(e);
			throw interruptedException;
		}
		if (rethrownInvocationTargetException.get() != null) {
			throw rethrownInvocationTargetException.get();
		}
		if (rethrownInterruptedException.get() != null) {
			throw rethrownInterruptedException.get();
		}
```

Reply:
That's actually the way I imagined it: the 2 `if`-statements at the end are also "a `switch` in disguise".
Or am I missing something?

Reply:
I think I don't get your point. Where do you see a switch? The method throws exactly those kinds of exceptions that you can assign within the runnable. There is no chance the runnable assigns a type of exception that is not rethrown in the method (unlike when having a generic `Exception` variable from which only specific types are extracted).

Reply:
What I mean with "a `switch` in disguise" is that when using several variables, you still need to exhaust all possibilities and query all variables for `null`.

If `SomeNewCheckedException` comes along then you still need to add this at the end:

```java
if (someNewCheckedException.get() != null) {
	throw someNewCheckedException.get();
}
```

This is the typical place were one does a "copy-paste mistake" that may end up like...

```java
if (someNewCheckedException.get() != null) {
	throw rethrownInterruptedException.get();
}
```
One ends up having a new variable and a new exclusive `catch`-clause (just more code to maintain) without really solving the problem of having to exhaust every possibility.

Reply:
> If `SomeNewCheckedException` comes along then you still need to add this at the end:

That's exactly the point. You _must_ do this. When just having an Exception-typed variable, you are not forced to add code for another exception, which results in that exception being swallowed. This is a risk and will eventually lead to a bug, like in almost every case where `instanceof` is applied without providing handling for every type.

As mentioned before: You do not necessarily need to use my proposal, just choose any option that ensures that a complete switch is performed on the possible types.